File: src/html/heatmap.js

const heatmap_data___id__ = __data__;

let heatmap_filters___id__ = {x: "", y: ""};


function render_heatmap___id__() {
    const data = heatmap_data___id__;
    const num_colors = __num_colors__;
    
    let index_x = [], index_y = [];
    for (let x=0; x<data.labels_x.length; ++x) {
        index_x.push(x);
    }
    for (let y=0; y<data.labels_y.length; ++y) {
        index_y.push(y);
    }
    let min_val = 0., max_val = 0.;
    for (const y of index_y) {
        const row = data.matrix[y];
        for (const x of index_x) {
            let value = row[x];
            min_val = Math.min(min_val, value);
            max_val = Math.max(max_val, value);
        }
    }
    const color_factor = min_val === max_val ? 0. : (num_colors) / (max_val - min_val);

    let html = "";

    html += `<div class="hmlabelv"></div>`;
    for (const x of index_x) {
        const label = data.labels_x[x];
        html += `<div class="hmlabelv" title="${label}">${label}</div>`;
    }

    for (const y of index_y) {
        const row = data.matrix[y];

        const label = data.labels_y[y];
        html += `<div class="hmlabel" title="${label}">${label}</div>`;

        for (const x of index_x) {
            const
                value = row[x],
                col = Math.min(__num_colors__ - 1, Math.floor((value - min_val) * color_factor));

            const title = `${data.labels_x[x]} / ${data.labels_y[y]}: ${value}`;
            html += `<div class="hmc v${col}" title="${title}"></div>`;
        }
    }

    html += `<div class="hmlabelvb"></div>`;
    for (const x of index_x) {
        const label = data.labels_x[x];
        html += `<div class="hmlabelvb" title="${label}">${label}</div>`;
    }

    const elem = document.querySelector(".heatmap-__id__");
    elem.style = `grid-template-columns: __label_width__ repeat(${index_x.length}, 1fr);`;
    elem.innerHTML = html;
}

render_heatmap___id__();

File: src/html/heatmap.py
import pathlib
import json
import secrets
from typing import Union, Optional, List

import pandas as pd

import plotly.express as px


_JS_SOURCE = pathlib.Path(__file__).resolve().parent.joinpath("heatmap.js").read_text()


def html_heatmap(
        matrix: Union[List[List], pd.DataFrame],
        labels_x: Optional[List[str]] = None,
        labels_y: Optional[List[str]] = None,
        title: str = None,
        transpose: bool = False,
        display: bool = True,
        colors: Union[str, List[str]] = "Viridis",
        label_width: str = "10rem",
):
    if display:
        from IPython.display import display, HTML

    if not isinstance(matrix, pd.DataFrame):
        matrix = pd.DataFrame(matrix)

    if not labels_x:
        labels_x = matrix.columns
    if not labels_y:
        labels_y = matrix.index

    if transpose:
        matrix = matrix.transpose()
        labels_x, labels_y = labels_y, labels_x

    if isinstance(colors, str):
        colors = [c[1] for c in px.colors.PLOTLY_SCALES[colors]]

    ID = secrets.token_hex(8)

    html = """<style>
        .heatmap-%(ID)s {
            display: grid;
        }
        .heatmap-%(ID)s .hmlabel,
        .heatmap-%(ID)s .hmlabelv,
        .heatmap-%(ID)s .hmlabelvb {
            text-align: right;
            padding-right: .3rem;
            overflow: hidden;
        }
        .heatmap-%(ID)s .hmlabelv,
        .heatmap-%(ID)s .hmlabelvb {
            writing-mode: vertical-lr;
            height: %(label_width)s;
            padding: 0;
            padding-bottom: .3rem;
        }
        .heatmap-%(ID)s .hmlabelvb {
            text-align: left;
            padding: 0;
            padding-top: .3rem;
        }
        .heatmap-%(ID)s .hmc {
            padding-bottom: 100%%;
            border: 1px solid black;
        }
    """ % {
        "ID": ID,
        "width": matrix.shape[1],
        "label_width": label_width,
    }
    for i, c in enumerate(colors):
        html += f""".heatmap-{ID} .hmc.v{i} {{ background: {c} }}"""
    html += """</style>"""

    if title:
        html += f"<h3>{title}</h3>"

    html += f"""<div class="heatmap-{ID}"></div>"""

    script_context = {
        "id": ID,
        "num_colors": len(colors),
        "label_width": label_width,
        "data": json.dumps({
            "matrix": matrix.values.tolist(),
            "labels_x": [str(l) for l in labels_x],
            "labels_y": [str(l) for l in labels_y],
        }),
    }
    script = _JS_SOURCE
    for key, value in script_context.items():
        script = script.replace(f"__{key}__", str(value))

    html += f"""<script type="text/javascript">{script}</script>"""

    html = "\n".join(line.strip() for line in html.splitlines())

    if display:
        display(HTML(html))
    else:
        return html


if __name__ == "__main__":
    import random

    def random_name():
        return "".join(
            random.choice(["ku", "ka", "bo", "ba", "su", "la", "to", "mi", "no", "ha"])
            for _ in range(random.randrange(3, 9))
        )

    rows = []
    value = 0
    for y in range(20):
        row = []
        for x in range(60):
            row.append(value)
            value += 1
        rows.append(row)

    labels_x = [random_name() for _ in rows[0]]
    labels_y = [random_name() for _ in rows]

    heatmap_html = html_heatmap(
        rows,
        labels_x=labels_x,
        labels_y=labels_y,
        display=False,
        #colors=["red", "green", "blue"],
    )

    html = f"""
    <!DOCTYPE html>
    <html>
    <body>
        <h3>heatmap test</h3>
        {heatmap_html}
    </body>
    </html>
    """
    with open("heatmap-test.html", "w") as fp:
        fp.write(html)
